File: examples/kotlin-kotest/src/test/kotlin/examples/library.steps.kt
@file:JvmName("LibrarySteps")

package examples

import com.oselvar.varkt.defineState
import com.oselvar.varkt.sensor
import com.oselvar.varkt.stimulus
import java.time.LocalDate
import java.util.Locale

data class LibraryCtx(
    val loans: List<Loan> = emptyList(),
    val feePence: Int = 0,
    val granted: Boolean = false,
)

private val MONTHS =
    listOf(
        "January",
        "February",
        "March",
        "April",
        "May",
        "June",
        "July",
        "August",
        "September",
        "October",
        "November",
        "December",
    )

/** June 6th → LocalDate 2026-06-06 (the spec's year is 2026). */
private fun toDate(raw: String): LocalDate {
    val (month, day) = raw.split(" ")
    return LocalDate.of(2026, MONTHS.indexOf(month) + 1, day.dropLast(2).toInt())
}

/** £2.50 and 50p, both as pence. */
private fun toPence(raw: String): Int =
    if (raw.startsWith("£")) Math.round(raw.drop(1).toDouble() * 100).toInt()
    else raw.dropLast(1).toInt()

val librarySteps =
    defineState(::LibraryCtx) {
        parameterType(
            "date",
            Regex(
                "(?:January|February|March|April|May|June" +
                    "|July|August|September|October|November|December)" +
                    " \\d{1,2}(?:st|nd|rd|th)"
            ),
        ) { groups ->
            toDate(groups[0])
        }
        parameterType(
            "money",
            Regex("£\\d+(?:\\.\\d{2})?|\\d+p"),
            // The inverse: mismatches render as £2.60 / 50p, not a bare pence int.
            format = { pence ->
                if (pence < 100) "${pence}p" else "£%.2f".format(Locale.ROOT, pence / 100.0)
            },
        ) { groups ->
            toPence(groups[0])
        }
        // Emphasis (*Emma*) is stripped before matching, so a title is a
        // Title Case run in the plain prose.
        parameterType("title", Regex("[A-Z][a-z]+(?: [A-Z][a-z]+)*")) { groups -> groups[0] }

        stimulus("borrowed {title}, due back on {date}") { title: String, due: LocalDate ->
            copy(loans = loans + Loan(title, due))
        }
        stimulus("returns it on {date}") { returnedOn: LocalDate ->
            copy(feePence = loans.sumOf { loan -> lateFee(loan, returnedOn) })
        }
        sensor("owes a {money} late fee") { _: Int -> feePence }
        sensor("{money} for each day overdue") { _: Int -> FEE_PENCE_PER_DAY }
        stimulus("asks to borrow {title} on {date}") { _: String, on: LocalDate ->
            copy(granted = mayBorrow(loans, on))
        }
        sensor("the library refuses") {
            if (granted) throw AssertionError("expected the library to refuse")
            null
        }
        sensor("the library agrees") {
            if (!granted) throw AssertionError("expected the library to agree")
            null
        }
    }

File: examples/go-gotest/library.steps.go
package example

import (
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/varar-dev/varar/go/varar"
)

// dateLayout is the document's date notation — "June 1, 2026" — in Go's
// reference-time spelling.
const dateLayout = "January 2, 2006"

func registerLibrary(s *varar.Steps[Ctx]) {
	s.Param("date", `[A-Z][a-z]+ \d{1,2}, \d{4}`,
		func(g []string) time.Time {
			t, err := time.Parse(dateLayout, g[0])
			if err != nil {
				panic(err)
			}
			return t
		},
		func(d time.Time) (string, bool) { return d.Format(dateLayout), true })

	s.Param("money", `£\d+(?:\.\d+)?|\d+p`,
		func(g []string) Money {
			raw := g[0]
			if pence, ok := strings.CutSuffix(raw, "p"); ok {
				return GBP(mustFloat(pence) / 100)
			}
			return GBP(mustFloat(strings.TrimPrefix(raw, "£")))
		},
		func(m Money) (string, bool) {
			if m.Value < 1 {
				return fmt.Sprintf("%dp", int64(math.Round(m.Value*100))), true
			}
			return fmt.Sprintf("£%.2f", m.Value), true
		})

	s.Stimulus("borrowed {emph}, due back on {date}", func(ctx Ctx, title string, due time.Time) (Ctx, error) {
		ctx.Loans = append(append([]Loan{}, ctx.Loans...), Loan{Title: title, Due: due})
		return ctx, nil
	})

	s.Stimulus("returns it on {date}", func(ctx Ctx, returned time.Time) (Ctx, error) {
		fee := GBP(0)
		for _, loan := range ctx.Loans {
			sum, err := AddMoney(fee, LateFee(loan, returned))
			if err != nil {
				return ctx, err
			}
			fee = sum
		}
		ctx.Fee = fee
		return ctx, nil
	})

	s.Sensor("owes a {money} late fee", func(ctx Ctx, expected Money) (Money, error) {
		return ctx.Fee, nil
	})

	s.Sensor("{money} for each day overdue", func(ctx Ctx, expected Money) (Money, error) {
		return FeePerDay, nil
	})

	s.Stimulus("asks to borrow {emph} on {date}", func(ctx Ctx, title string, on time.Time) (Ctx, error) {
		ctx.Granted = MayBorrow(ctx.Loans, on)
		return ctx, nil
	})

	s.Sensor("the library refuses", func(ctx Ctx) error {
		if ctx.Granted {
			return errors.New("expected the library to refuse")
		}
		return nil
	})

	s.Sensor("the library agrees", func(ctx Ctx) error {
		if !ctx.Granted {
			return errors.New("expected the library to agree")
		}
		return nil
	})
}

func mustFloat(s string) float64 {
	f, err := strconv.ParseFloat(s, 64)
	if err != nil {
		panic("not a number: " + s)
	}
	return f
}
